sysvabi64/sysvabi64.rst

```diff
@@ -0,0 +1,1076 @@
+addressing to access data which is all within +/- 4GiB of the code.
+
+.. code-block:: asm
+
+  foo:
+    adrp    x0, dst                      // R_AARCH64_ADR_PREL_PG_HI21   dst
+    add     x1, x0, :lo12:dst            // R_AARCH64_LDST64_ABS_LO12_NC dst
+    adrp    x2, src                      // R_AARCH64_ADR_PREL_PG_HI21   src
+    ldrb    w2, [x2, #:lo12:src]         // R_AARCH64_LDST64_ABS_LO12_NC src
+    strb    w2, [x0, #:lo12:dst]         // R_AARCH64_LDST64_ABS_LO12_NC src
+    adrp    x0, ptr                      // R_AARCH64_ADR_PREL_PG_HI21   ptr
+    str     x1, [x0, #:lo12:ptr]         // R_AARCH64_LDST64_ABS_LO12_NC ptr
+    ret
+
+  bar:
+    adrp    x1, .LANCHOR0                // R_AARCH64_ADR_PREL_PG_HI21   .LANCHOR0
+    add     x0, x1, :lo12:.LANCHOR0      // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0
+    ldrb    w2, [x0, 1]
+    strb    w2, [x1, #:lo12:.LANCHOR0]   // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0
+    str     x0, [x0, 8]
+    ret
+
+  baz:
+    adrp    x0, lbdst                    // R_AARCH64_ADR_PREL_PG_HI21   lbdst
+    add     x1, x0, :lo12:lbdst          // R_AARCH64_LDST64_ABS_LO12_NC lbdst
+    adrp    x2, lbsrc                    // R_AARCH64_ADR_PREL_PG_HI21   lbsrc
+    ldrb    w2, [x2, #:lo12:lbsrc]       // R_AARCH64_LDST64_ABS_LO12_NC lbsrc
+    strb    w2, [x0, #:lo12:lbdst]       // R_AARCH64_LDST64_ABS_LO12_NC lbdst
+    adrp    x0, .LANCHOR0+8              // R_AARCH64_ADR_PREL_PG_HI21   .LANCHOR0 + 8
+    str     x1, [x0, #:lo12:.LANCHOR0+8] // R_AARCH64_LDST64_ABS_LO12_NC .LANCHOR0 + 8
+    ret
+
+  getweakaddr:
+    adrp    x0, .LC0                     // R_AARCH64_ADR_PREL_PG_HI21   .LC0
+    ldr     x0, [x0, #:lo12:.LC0]        // R_AARCH64_LDST64_ABS_LO12_NC .LC0
+    ret
+    .align 3
+  .LC0:
+    .xword  weakref                      // R_AARCH64_ABS64 weakref
+
+  .bss
+  .LANCHOR0
+  ldst:
+    .zero   1
+    .zero   7
+  lbdst:
+    .zero   65536
+
+Large Code Model
+^^^^^^^^^^^^^^^^
+
+The large model makes no assumptions about the size of the static code
+and data segments, and therefore uses absolute 64-bit addresses in all
+cases. As well as massive GNU/Linux executables, the large code model
+would also be suitable for bare-metal programs where a linker script
+causes sections to be mapped sparsely within the virtual address
+space, or where a high-level code references absolute symbols defined
+in a linker script.
+
+.. code-block:: asm
+
+  foo:
+    adrp    x0, .LC0              // R_AARCH64_ADR_PREL_PG_HI21   .LC0
+    ldr     x0, [x0, #:lo12:.LC0] // R_AARCH64_LDST64_ABS_LO12_NC .LC0
+    adrp    x1, .LC1              // R_AARCH64_ADR_PREL_PG_HI21   .LC1
+    ldr     x1, [x1, #:lo12:.LC1] // R_AARCH64_LDST64_ABS_LO12_NC .LC1
+    ldrb    w1, [x1]
+    strb    w1, [x0]
+    adrp    x1, .LC2              // R_AARCH64_ADR_PREL_PG_HI21   .LC2
+    ldr     x1, [x1, #:lo12:.LC2] // R_AARCH64_LDST64_ABS_LO12_NC .LC2
+    str     x0, [x1]
+    ret
+    .align 3
+  .LC0:
```

Comment:
I've added an additional example showing clang's output which uses 4 movs/movk rather than use literal pools. As mentioned before I'd like to keep the examples in the first revision to be backed up by an implementation.

Reply:
Unfortunately both GCC and LLVM implementations are awful and inefficient...